src/swupdd-main.c

```diff
@@ -22,751 +18,765 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
  * 02110-1301 USA
  *
+ * Contact: Dmitry Rozhkov <dmitry.rozhkov@intel.com>
+ *
  */
 
-#include <stdio.h>
+#define _GNU_SOURCE
+
 #include <stdlib.h>
-#include <string.h>
-#include <signal.h>
 #include <unistd.h>
-#include <sys/types.h>
-
-#include <glib.h>
-#include <glib-unix.h>
-#include <gio/gio.h>
-
-#include "swupdd-interface.h"
+#include <stdio.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <limits.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <systemd/sd-bus.h>
+#include <systemd/sd-daemon.h>
+
+#include "log.h"
+#include "list.h"
 
 #define SWUPD_CLIENT    "swupd"
-
-
-typedef struct _daemon_state
-{
-    guint busid;
-    GDBusConnection *conn;
-    ClrSoftwareUpdate *upif;
-    const gchar *url;
-    const gchar *method;
-    GSubprocess *process;
+#define TIMEOUT_EXIT_SEC 30
+
+typedef enum {
+	METHOD_NOTSET = 0,
+	METHOD_CHECK_UPDATE,
+	METHOD_UPDATE,
+	METHOD_VERIFY,
+	METHOD_BUNDLE_ADD,
+	METHOD_BUNDLE_REMOVE
+} method_t;
+
+typedef struct _daemon_state {
+	sd_bus *bus;
+	pid_t child;
+	method_t method;
 } daemon_state_t;
 
-
-static const gchar *_empty_strv[] = { NULL, };
-
-
-static gchar ** _list_to_strv (GList *list)
+static const char * const _method_str_map[] = {
+	NULL,
+	"checkUpdate",
+	"update",
+	"verify",
+	"bundleAdd",
+	"bundleRemove"
+};
+
+static const char * const _method_opt_map[] = {
+	NULL,
+	"check-update",
+	"update",
+	"verify",
+	"bundle-add",
+	"bundle-remove"
+};
+
+static char **list_to_strv(struct list *strlist)
 {
-    gchar **strv;
-    gchar **temp;
-
-    strv = g_new0 (gchar *, g_list_length (list) + 1);
-
-    temp = strv;
-    while (list)
-    {
-        *temp = list->data;
-        temp++;
-        list = g_list_next (list);
-    }
-    return strv;
+	char **strv;
+	char **temp;
+
+	strv = (char **)malloc((list_len(strlist) + 1) * sizeof(char *));
+	memset(strv, 0x00, (list_len(strlist) + 1) * sizeof(char *));
```

Comment:
minor, buy using `calloc(list_len(strlist) + 1, sizeof(char *))` would be simpler.